### PR TITLE
Implement zeroing of gradients

### LINKED ISSUE #336: Implement zeroing of gradients
Following @jatkinson1000's question about what happens when we run backpropagation on different tensors in https://github.com/Cambridge-ICCS/FTorch/pull/286#discussion_r1998305043, I did some digging.

I think this PyTorch code encapsulates things nicely (print outputs showed as comments):
```python
import torch

a = torch.tensor([1.0], requires_grad=True)
b = torch.tensor([2.0], requires_grad=True)

Q = a * b
print(f"Q = a * b = {Q}")
# Q = a * b = tensor([2.], grad_fn=<MulBackward0>)

P = a * b
print(f"P = a * b = {P}")
# P = a * b = tensor([2.], grad_fn=<MulBackward0>)

Q.backward(torch.ones((1,)), retain_graph=True)
print(f"dQda = {a.grad} ?= b = {b}")
print(f"dQdb = {b.grad} ?= a = {a}")
# dQda = tensor([2.]) ?= b = tensor([2.], requires_grad=True)
# dQdb = tensor([1.]) ?= a = tensor([1.], requires_grad=True)

P.backward(torch.ones((1,)), retain_graph=True)
print(f"dPda = {a.grad} ?= b = {b}")
print(f"dPdb = {b.grad} ?= a = {a}")
# dPda = tensor([4.]) ?= b = tensor([2.], requires_grad=True)
# dPdb = tensor([2.]) ?= a = tensor([1.], requires_grad=True)

R = a * b
print(f"R = a * b = {R}")
# R = a * b = tensor([2.], grad_fn=<MulBackward0>)

R.backward(torch.ones((1,)), retain_graph=True)
print(f"dRda = {a.grad} ?= b = {b}")
print(f"dRdb = {b.grad} ?= a = {a}")
# dRda = tensor([6.]) ?= b = tensor([2.], requires_grad=True)
# dRdb = tensor([3.]) ?= a = tensor([1.], requires_grad=True)

a.grad.zero_()
b.grad.zero_()

S = a * b
print(f"S = a * b = {S}")
# S = a * b = tensor([2.], grad_fn=<MulBackward0>)

S.backward(torch.ones((1,)), retain_graph=True)
print(f"dSda = {a.grad} ?= b = {b}")
print(f"dSdb = {b.grad} ?= a = {a}")
# dSda = tensor([2.]) ?= b = tensor([2.], requires_grad=True)
# dSdb = tensor([1.]) ?= a = tensor([1.], requires_grad=True)
```

In order for the backpropagation to give the correct gradients, we need to zero the gradients beforehand.

### PR DESCRIPTION
**Pull request by jwallwork23**
Closes #336.

This PR adds a subroutine (and corresponding class method) for zeroing the gradients associated with the tensor.

The caveat that had been tripping me up until now is that you need to call `torch_tensor_get_gradient` after zeroing the gradients. It'd be nice if we could find a way round this, but I guess it's okay for now so long as it's documented.

### CODE REVIEW

File: pages/autograd.md [hunk 1]
```diff
@@ -71,15 +71,45 @@ to `a` and/or `b`. To do this, we can use the `torch_tensor_get_gradient`
...
+call torch_tensor_backward(P)
+
+! ...
+```
+
+#### Extracting gradients
+
+Note that `torch_tensor_get_gradient` must be called after every call to
+`torch_tensor_backward` or `torch_tensor_zero_grad`, even if the gradient for
+the same tensor is being extracted into the same array.
```
Comment (jatkinson1000) [inline pages/autograd.md:111]: Might be useful to add a short "because".
Related to changing pointer or data location by these calls on the C++ side?

File: src/ctorch.h [hunk 2]
```diff
@@ -251,6 +251,12 @@ EXPORT_C void torch_tensor_power_float(torch_tensor_t output,
 // --- Functions related to automatic differentiation functionality for tensors
 // =============================================================================
 
+/**
+ * Function to reset the gradient values of a Torch Tensor to zero
+ * @param Torch Tensor to zero the graident values of
```
Comment (jatkinson1000) [inline src/ctorch.h:256]: small typo
```suggestion
 * @param Torch Tensor to zero the gradient values of
```

File: src/ftorch.fypp [hunk 3]
```diff
@@ -840,6 +841,23 @@ contains
   ! --- Procedures related to automatic differentation functionality for tensors
   ! ============================================================================
 
+  !> Resets a tensor's gradient to zero.
+  subroutine torch_tensor_zero_grad(tensor)
+    use, intrinsic :: iso_c_binding, only : c_associated
+    class(torch_tensor), intent(inout) :: tensor
```
Comment (jatkinson1000) [inline src/ftorch.fypp:847]:
```suggestion
    class(torch_tensor), intent(inout) :: tensor  !! Tensor to zero the gradient of
```

**Review by jatkinson1000 — APPROVED** [review]
All generally seems sensible @jwallwork23

"Request changes" seems a bit heavy, but small typo correction and addition of inline FORD docs for you.
Also a small docs addition to explain why users need to get_gradient after each recompute may be useful if you agree.

File: src/ctorch.h (hunk 2 above)
Comment (jwallwork23) [reply src/ctorch.h:256]: Thanks, I always make this typo!

Comment (jatkinson1000) [reply src/ctorch.h]: 😂 Yeah, I have certain words that get me every time too!

File: pages/autograd.md (hunk 1 above)
Comment (jwallwork23) [reply pages/autograd.md:121]: I think it's because of the way we use `std::move`. I wonder if we can avoid this.

Comment (jatkinson1000) [reply pages/autograd.md]: Ah, yes, I actually have this in a PR that I am yet to push changes for.
